src/main/java/com/example/autokis/parallel/service/ParallelParsingService.java

```diff
@@ -0,0 +1,79 @@
+package com.example.autokis.parallel.service;
+
+
+import com.example.autokis.parallel.ParallelParsingTask;
+import com.example.autokis.parallel.config.ParallelParsingConfig;
+import com.example.autokis.parser.WebStoreProductParser;
+import com.example.autokis.parser.model.Category;
+import com.example.autokis.parser.model.Product;
+import com.example.autokis.parser.provider.model.CategoryContext;
+import com.example.autokis.parser.provider.model.CategoryToProductUrl;
+import com.google.common.collect.Lists;
+import lombok.AllArgsConstructor;
+import org.springframework.stereotype.Service;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+@Service
+@AllArgsConstructor
+public class ParallelParsingService {
+    private final ParallelParsingConfig parallelParsingConfig;
+
+    public List<Product> parseProductsInParallel(List<CategoryContext> categoryContexts, WebStoreProductParser webStoreProductParser) {
+        try {
+            int threadNumber = Integer.parseInt(parallelParsingConfig.getThreadNumber());
+            List<ParallelParsingTask> tasksToExecute = new ArrayList<>();
+            List<CategoryToProductUrl> categoryToProductUrls = getProductToCategoryMappingFromContext(categoryContexts);
+            List<List<CategoryToProductUrl>> threadPartitions = Lists.partition(categoryToProductUrls, categoryToProductUrls.size() / threadNumber);
```

Comment:
Here we can calculate the number of partitions like this:
`Math.ceil(categoryToProductUrls.size() / (double) threadNumber)`
This guarantees us that we handle all the products